compiler/passes/errorHandling.cpp

```diff
@@ -396,3 +422,148 @@ void lowerErrorHandling() {
     fn->accept(&visitor);
   }
 }
+
+namespace {
+
+class CanThrowVisitor : public AstVisitorTraverse {
+
+public:
+  CanThrowVisitor       (bool inThrowingFn, bool makeCompileErrors);
+
+  virtual bool enterTryStmt  (TryStmt*   node);
+  virtual void exitTryStmt   (TryStmt*   node);
+  virtual bool enterCatchStmt(CatchStmt*   node);
+  virtual void exitCatchStmt (CatchStmt*   node);
+  virtual bool enterCallExpr (CallExpr*  node);
+
+  bool throws() { return canThrow; }
+
+private:
+  int  tryDepth;
+  bool canThrow;
+  bool errors;
+  bool fnCanThrow; // only used for error checking
+
+  bool   catchesNotExhaustive(TryStmt* tryStmt);
+};
+
+CanThrowVisitor::CanThrowVisitor(bool inThrowingFn, bool makeCompileErrors) {
+  tryDepth = 0;
+  canThrow = false;
+  errors = makeCompileErrors;
+  fnCanThrow = inThrowingFn;
+}
+
+bool CanThrowVisitor::enterTryStmt(TryStmt* node) {
+  tryDepth++;
+
+  return true;
+}
+
+void CanThrowVisitor::exitTryStmt(TryStmt* node) {
+  tryDepth--;
+
+  // is it an exhaustive catch?
+
+  bool nonExhaustive = catchesNotExhaustive(node);
+
+  if (node->tryBang()) {
+    canThrow = false;
+  } else {
+    canThrow = nonExhaustive;
+    if (errors && tryDepth==0 && nonExhaustive && !fnCanThrow) {
+      USR_FATAL_CONT(node, "try without a catchall in a non-throwing function");
+    }
+  }
+}
+
+bool CanThrowVisitor::catchesNotExhaustive(TryStmt* tryStmt) {
+
+  bool       hasCatchAll = false;
+
+  for_alist(c, tryStmt->_catches) {
+    if (errors && hasCatchAll)
+      USR_FATAL_CONT(c->prev, "catchall placed before the end of a catch list");
+
+    SET_LINENO(c);
+
+    CatchStmt* catchStmt = toCatchStmt(c);
+    DefExpr*   catchDef  = catchStmt->expr();
+
+    // catchall
+    if (catchDef == NULL) {
+      hasCatchAll = true;
+    } else {
+      VarSymbol* errSym  = toVarSymbol(catchDef->sym);
+      Type*      errType = errSym->type;
+
+      // named catchall
+      if (errType == dtError) {
+        hasCatchAll = true;
+      }
+    }
+  }
+
+  return !hasCatchAll;
+}
+
+bool CanThrowVisitor::enterCatchStmt(CatchStmt* node) {
+  return true;
+}
+
+void CanThrowVisitor::exitCatchStmt(CatchStmt* node) {
+}
```

Comment:
You can remove `enter` and `exitCatchStmt`, as these are the same as the `AstVisitorTraverse` implementations.